cirq-core/cirq/optimizers/transformer_primitives.py

```diff
@@ -129,6 +130,118 @@ def map_operations_and_unroll(
     return unroll_circuit_op(map_operations(circuit, map_func))
 
 
+def merge_operations(
+    circuit: CIRCUIT_TYPE,
+    merge_func: Callable[[ops.Operation, ops.Operation], Optional[ops.Operation]],
+) -> CIRCUIT_TYPE:
+    """Merges operations in a circuit by calling `merge_func` iteratively on operations.
+
+    Two operations op1 and op2 are merge-able if
+        - There is no other operations between op1 and op2 in the circuit
+        - is_subset(op1.qubits, op2.qubits) or is_subset(op2.qubits, op1.qubits)
+
+    The `merge_func` is a callable which, given two merge-able operations
+    op1 and op2, decides whether they should be merged into a single operation
+    or not. If not, it should return None, else it should return the single merged
+    operations `op`.
+
+    If op1 and op2 are merged, both op1 and op2 are deleted from the circuit and
+    the resulting `merged_op` is inserted at the index corresponding to the larger
+    of op1/op2. If both op1 and op2 act on the same number of qubits, `merged_op` is
+    inserted in the smaller moment index to minimize circuit depth.
+
+    The number of calls to `merge_func` is O(N), where N = Total no. of operations, because:
+        - Every time the `merge_func` returns a new operation, the number of operations in the
+            circuit reduce by 1 and hence this can happen at most O(N) times
+        - Every time the `merge_func` returns None, the current operation is inserted into the
+            frontier and we go on to process the next operation, which can also happen at-most
+            O(N) times.
+
+    Args:
+        circuit: Input circuit to apply the transformations on. The input circuit is not mutated.
+        merge_func: Callable to determine whether two merge-able operations in the circuit should
+            be merged. If the operations can be merged, the callable should return the merged
+            operation, else None.
+
+    Returns:
+        Copy of input circuit with merged operations.
+
+    Raises:
+        ValueError if the merged operation acts on new qubits outside the set of qubits
+            corresponding to the original operations to be merged.
+    """
+
+    def apply_merge_func(op1: ops.Operation, op2: ops.Operation) -> Optional[ops.Operation]:
+        new_op = merge_func(op1, op2)
+        qubit_set = frozenset(op1.qubits + op2.qubits)
+        if new_op is not None and not qubit_set.issuperset(new_op.qubits):
+            raise ValueError(
+                f"Merged operation {new_op} must act on a subset of qubits of "
+                f"original operations {op1} and {op2}"
+            )
+        return new_op
+
+    ret_circuit = circuits.Circuit()
+    for current_moment in circuit:
+        new_moment = ops.Moment()
+        for op in current_moment:
+            op_qs = set(op.qubits)
+            idx = ret_circuit.prev_moment_operating_on(tuple(op_qs))
```

Comment:
I agree that we can speed this up by explicitly maintaining the frontier but I'd propose that we can create a new issue to track this performance improvement and let this get checked in for now. This would unblock migration of existing transformers to use the new primitives and the performance improvement would be an internal implementation change.